core/src/test/java/org/pentaho/platform/util/VersionInfoTest.java

```diff
@@ -0,0 +1,263 @@
+    version = "10.3a.123.007.87590.foe";
+    versionClean = "10.3a.123.007.87590";
+    info.setVersion( version );
+    doTest( info, false, null, null, "10", "3a", "123", "007", "87590", versionClean );
+
+    info = new VersionInfo();
+    version = "10.3a.123.007.87590.foe.foo-abc.xyz";
+    versionClean = "10.3a.123.007.87590";
+    info.setVersion( version );
+    doTest( info, false, null, null, "10", "3a", "123", "007", "87590", versionClean );
+  }
+
+  /**
+   * Tests that when partial version values are provided (with missing pieces in the beginning, middle and end),
+   * the individual pieces are set as expected.
+   */
+  @Test
+  public void testSetVersionMissingFields() {
+    VersionInfo info;
+    String version, versionClean;
+
+    // when the first character is the delimiter (.) the major version should be empty
+    info = new VersionInfo();
+    version = ".3a.123.007.";
+    versionClean = ".3a.123.007";
+    info.setVersion( version );
+    doTest( info, false, null, null, "",  "3a", "123", "007", null, versionClean );
+
+    // when the first character is the delimiter (-) the major version should be empty
+    info = new VersionInfo();
+    version = "-3a.123.007-";
+    versionClean = ".3a.123.007";
+    info.setVersion( version );
+    doTest( info, false, null, null, "", "3a", "123", "007", null, versionClean );
+
+    // consecutive delimiter characters yield empty values where appropriate
+    info = new VersionInfo();
+    version = "10..123..87590";
+    versionClean = "10..123..87590";
+    info.setVersion( version );
+    doTest( info, false, null, null, "10", "", "123", "", "87590", versionClean );
+
+    info = new VersionInfo();
+    version = "10.3a.-.87590";
+    versionClean = "10.3a...87590";
+    info.setVersion( version );
+    doTest( info, false, null, null, "10", "3a", "", "", "87590", versionClean );
+  }
+
+  private static void doTest( final VersionInfo info, final boolean expectedFromManifest, final String
+    expectedProductID, final String expectedTitle, final String expectedVersionMajor, final String
+    expectedVersionMinor, final String expectedVersionRelease, final String expectedVersionMilestone, final String
+    expectedVersionBuild, final String expectedVersionNumber ) {
+
+    Assert.assertEquals( expectedFromManifest, info.isFromManifest() );
+    Assert.assertEquals( expectedProductID, info.getProductID() );
+    Assert.assertEquals( expectedTitle, info.getTitle() );
+    Assert.assertEquals( expectedVersionMajor, info.getVersionMajor() );
+    Assert.assertEquals( expectedVersionMinor, info.getVersionMinor() );
+    Assert.assertEquals( expectedVersionRelease, info.getVersionRelease() );
+    Assert.assertEquals( expectedVersionMilestone, info.getVersionMilestone() );
+    Assert.assertEquals( expectedVersionBuild, info.getVersionBuild() );
+    Assert.assertEquals( expectedVersionNumber, info.getVersionNumber() );
+
+    final StringBuffer buffer = new StringBuffer();
```

Comment:
Do we want/need to have a test on to versionInfo.toString()? Not sure what the answer is... it depends if this method must return a string in an standard format. I.e. do we have a contract on the format of toString()? If not, I would recommend removing this test.